### PR TITLE
automate the ironbank generation

### PR DESCRIPTION
**Pull request by v1v**
## Motivation/summary

Automate the docker context generation for the IronBank releases, this will allow us to move away from creating those docker context manually in a different repository, since the Platform Release team already provide the automation.

## Implementation details

A similar approach was done for [Kibana](https://github.com/elastic/kibana/pull/89933) and [Elasticsearch](https://github.com/elastic/elasticsearch/pull/64336)

I initially enabled to build and save the docker image but it failed with some auth issues:

```
mage ironbank
>> Building docker images for IronBank
[+] Building 3.0s (3/3) FINISHED                                         
 => [internal] load build definition from Dockerfile                0.0s
 => => transferring dockerfile: 3.30kB                              0.0s
 => [internal] load .dockerignore                                   0.0s
 => => transferring context: 2B                                     0.0s
 => ERROR [internal] load metadata for registry1.dsop.io/ironbank/  2.9s
------
 > [internal] load metadata for registry1.dsop.io/ironbank/redhat/ubi/ubi8:8.6:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: unexpected status code [manifests 8.6]: 401 Unauthorized
Error: running "docker build -t docker.elastic.co/apm/apm-server-ironbank:8.4.0 build/ironbank" failed with exit code 1
failed to build docker
```

I don't know if other teams are actually building those docker.

## DoD build context content

As required the build context contains the following files:

- `Dockerfile` with specific requirements for DoD (registry args, no internet dependencies, healthcheck, ...)
- `hardening_manifest.yaml` with all required dependencies
- `LICENSE` specific for DoD
- `README.md` with specific content for DoD
- `config/` (optional) directory containing all config files to include into the Docker image
- `scripts/` directory containing all scripts to include into the Docker image (example: entrypoint...)

## Tasks

- [x] Exclude if not running for `x86_64`.
- [ ] Enable configuration in the DRA so it can be automated.
- [ ] Gather feedback from the teams.

## Test

image

This [job](https://apm-ci.elastic.co/job/apm-server/job/apm-server-package-mbp/job/PR-8537/) produces the artifacts.

## Further details

This could be potentially added to the Beats project, but I don't know what's the plan for the project and what's the priorities for the Platform Productivity team, but for now I think we could try to use this in the `apm-server` repo

### CODE REVIEW

**Comment by cachedout** [comment]
@v1v The implementation looks pretty sane. I am curious about this: `Enable configuration in DRA so it can be automated.` Does DRA have the capability to pick up an artifact and perform the rest of the needed automation?

File: packaging/ironbank/hardening_manifest.yaml.tmpl [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  mil.dso.ironbank.image.keywords: "apm,observabilty,o11y,oblt,performance,monitoring,elastic,elasticsearch,golang"
+  ## This value can be "opensource" or "commercial"
+  mil.dso.ironbank.image.type: "commercial"
+  ## Product the image belongs to for grouping multiple images
+  mil.dso.ironbank.product.name: "beats"
+
+# List of resources to make available to the offline build context
+resources:
+  - filename: "apm-server-{{ beat_version }}-linux-x86_64.tar.gz"
+    url: "<artifact_path>/apm-server-8.3.1-linux-x86_64.tar.gz"
```
Comment (v1v) [inline packaging/ironbank/hardening_manifest.yaml.tmpl:43]:
```suggestion
    url: "<artifact_path>/apm-server-{{ beat_version }}-linux-x86_64.tar.gz"
```

File: packaging/ironbank/hardening_manifest.yaml.tmpl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  ## Product the image belongs to for grouping multiple images
+  mil.dso.ironbank.product.name: "beats"
+
+# List of resources to make available to the offline build context
+resources:
+  - filename: "apm-server-{{ beat_version }}-linux-x86_64.tar.gz"
+    url: "<artifact_path>/apm-server-{{ beat_version }}-linux-x86_64.tar.gz"
+    validation:
+      type: "sha512"
+      value: "<insert SHA 512 here>"
```
Comment (v1v) [inline packaging/ironbank/hardening_manifest.yaml.tmpl:46]: This is the agreement with the Unified Release automation.

**Review by kuisathaverat — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Comment by cachedout** [comment]
@v1v I am not sure if apm-server needs to review this or not. I'm not sure if they're even aware of how the Ironbank stuff works. Perhaps @simitt knows?

**Review by marclop — COMMENTED** [review]
I don't know much about the ironbark generation, just a generic review on the errors returned

File: .gitignore [hunk 3]
```diff
@@ -53,3 +53,6 @@ terraform.tfstate.backup
 terraform.tfvars
 testing/benchmark/.envrc
 testing/benchmark/benchmark-result.txt
+
+# Ironbank
+apm-server-ironbank*
```
[the remark this answers is not in the document]
Comment (axw) [reply .gitignore:58]: Sorry, having trouble understanding this. What is the requirement? I'm not keen on build files being created in the top level directory. Why can't the `apm-server-ironbank-<version>-docker-build-context` directory be created under `build/`?

[the remark this answers is not in the document]
Comment (v1v) [reply .gitignore:58]: I tried initially that approach but the generated `apm-server-ironbank-<version>-docker-build-context.tar.gz` included the `build/...` folders. Unfortunately I could not find the way to `chdir` to `build` so the `tar` command could be created from that context instead.

image


I might need to look for how to solve this, though if you have some suggestions please let me know.

**Comment by cachedout** [comment]
@marclop Could you please give this one last look? Thanks.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (axw) [reply .gitignore:58]: @v1v should it even have `apm-server-ironbank-8.4.0-docker-build-context/` in the contents? This is what the 8.3.2 Elasticsearch IronBank build context tarball looks like:

```
$ tar tf elasticsearch-ironbank-8.3.2-docker-build-context.tar.gz
scripts/
scripts/docker-openjdk
scripts/docker-entrypoint.sh
scripts/elasticsearch.yml
README.md
LICENSE
hardening_manifest.yaml
Dockerfile
scripts/log4j2.properties
```

[the remark this answers is not in the document]
Comment (v1v) [reply .gitignore:58]: Good point, I thought I did review the folder layout for some other projects and it seems I messed up with the extra folder :/

```bash
tar -xvf kibana-ironbank-8.3.3-SNAPSHOT-docker-build-context.tar.gz
x Dockerfile
x LICENSE
x README.md
x bin/
x bin/kibana-docker
x config/
x config/kibana.yml
x hardening_manifest.yaml
```

[the remark this answers is not in the document]
Comment (v1v) [reply .gitignore:58]: I'd say `mage.Tar` cannot be used as is, I'll transition this PR to draft so I'll make the changes to support the `tar.gz` without any nested folders

File: .gitignore [hunk 4]
```diff
@@ -53,3 +53,4 @@ terraform.tfstate.backup
 terraform.tfvars
 testing/benchmark/.envrc
 testing/benchmark/benchmark-result.txt
+
```
Comment (axw) [inline .gitignore:56]:
```suggestion
```

**Review by axw — APPROVED** [review]
LGTM. As mentioned on Slack, I'm thinking about introducing our own Dockerfile, so I might end up making some changes. If we do, I'll ping you to find out how we can verify the changes.

**Pull request merged**